library/Faktory/Worker.hs

```diff
@@ -89,10 +87,7 @@ heartBeat client workerId = do
 
 fetchJob :: FromJSON args => Client -> Queue -> IO (Maybe (Job args))
 fetchJob client queue =
-  -- Treat an exception here like no Job. The Client will have already logged
-  -- the protocol error to stderr, and we should just keep chugging.
-  handleAny (const $ pure Nothing)
```

Comment:
@pbrisbin This is still bad behavior because it prevents the socket failure from appearing here.

Reply:
> Fetching was over eager about catching protocol errors. Instead of
> preventing failure from protocol and socket errors it should fail. This
> prevents bugs in the protocol or operational failures from being
> perceived as "no jobs"

Reply:
Yeah, I saw that. I was hoping for a little elaboration, or to find that the answer to

> I thought that #18 resolved the only concrete bad behavior we know of. Is that right?

Was no, making this the real fix.

I don't care enough to understand to be honest. I'm fine with this change; as long as the `WorkerHalt` behavior is corrected I'll approve the PR.

Reply:
#18 resolves issues of exceptions being silenced within the heartbeat, but if the heartbeat is removed then a failure could never bubble up because of the above, it would always be interpreted as `Nothing` instead of a socket exception. This fix just prevents a known issue from some day becoming problematic.

Reply:
I'm still not totally understanding, I'm sorry. If you're interested in helping me out here are my confusions, but feel free to not spend too much time.

>  but if the heartbeat is removed

What does this mean? How could that happen?

> it would always be interpreted as Nothing instead of a socket exception

Sure, but does that still lead to high CPU and ignoring of future Jobs? Or do subsequent fetches eventually stabilize back into a functioning consumer? The latter behavior was the intent behind the previous code.

Reply:
> >but if the heartbeat is removed
>
> What does this mean? How could that h

Without hearbeat throwing socket failures to the parent thread, this could never crash correctly. If that behavior is ever changed or a bug is reintroduced then we easily fall back into the previous behavior of always up, but not receiving jobs.

> Sure, but does that still lead to high CPU and ignoring of future Jobs? Or do subsequent fetches eventually stabilize back into a functioning consumer? The latter behavior was the intent behind the previous code

This isn't so much about high CPU as it is about correctly failing when a socket is lost or a protocol failure is introduce. Right now if a socket is lost `handleAny` will convert the exception into `Nothing` and `fetchJob` will continue to happily spin on a broken pipe. If faktory introduces a change to a protocol that breaks our parsing, `handleAny` will convert that to `Nothing` and `fetchJob` will continue to spin pretending it doesn't have any jobs to process. Instead we should blow up.

Reply:
OK, I think I understand now. Thanks for going back over it.

To confirm / recap:

**I asked**:

> I thought that #18 resolved the only concrete bad behavior we know of. Is that right? If so, which hypothetical bad behaviors are these changes aimed at?

**I now understand**:

> Yes that's right. The hypothetical bad behaviors are if that regresses down the line. If we also change this code, we're adding defense-in-depth with little downside.

**I asked**:

> Sure, but does that still lead to high CPU and ignoring of future Jobs? Or do subsequent fetches eventually stabilize back into a functioning consumer? The latter behavior was the intent behind the previous code.

**I now understand**:

> It won't stabilize back into a functioning consumer if it's a socket-lost exception, because nothing would reset the socket. For other exceptions it may.

I would personally rather keep the `handleAny` for protocol failures because I think a graceful ignore is more beneficial than it is harmful for that case. But I don't feel strongly, and the socket-lost case overshadows that anyway. So since there's no convenience way to handle the scenarios differently, crash away :+1:.

Reply:
What kind of exceptions would you expect from a protocol issue? We could handle those exceptions specifically. `handleAny` just ends up being dangerous because it can silence every possible exception.

Reply:
If I'm reading things correctly then protocol issues would surface in `Left` from
```
readReply :: IO ByteString -> IO (Either String (Maybe ByteString))
```

Which would then be converted to `Nothing` though
```
  case eByteString of
    Left err -> do
      settingsLogError err
      pure Nothing
```
So I believe protocol issues aren't being thrown, just `IOException`s that come off the socket.

Reply:
> What kind of exceptions would you expect from a protocol issue? We could handle those exceptions specifically.

"Exception" and "protocol issue" might both be incorrect and/or too vague. "Unexpected responses" might be more accurate. For example, we don't have any handling of the `SHUTDOWN` responses that are part of the wire protocol already. And we, by definition, have no support for responses that may be added in new server versions in the future.

My thinking is this: since this client has been built in sort of an as-needed / simplest-possible-thing way, I think it's in line with that approach to generally ignore un-recognized responses (and implicitly rely on downstream failures to happen if those responses actually mattered for some reason). Taking the crash approach just means we'll probably have to be more on top of adding support for more responses if we found they result in an unnecessary crash. It's just the classic permissive-vs-strict trade-off.

> So I believe protocol issues aren't being thrown, just IOExceptions that come off the socket.

Cool, that behavior SGTM then.